src/qtils/shared_ref.hpp

```diff
@@ -0,0 +1,234 @@
+     * @tparam U Other type
+     * @param other Another StrictSharedPtr
+     * @return true if this precedes other in ownership
+     */
+    template <typename U>
+    bool owner_before(const SharedRef<U> &other) const noexcept {
+      return ptr_.owner_before(other.ptr_);
+    }
+
+    /**
+     * @brief Compare ownership with a raw std::shared_ptr
+     * @tparam U Other type
+     * @param other Shared pointer
+     * @return true if this precedes other in ownership
+     */
+    template <typename U>
+    bool owner_before(const std::shared_ptr<U> &other) const noexcept {
+      return ptr_.owner_before(other);
+    }
+
+    /// @brief Swap with another SharedRef
+    void swap(SharedRef &other) noexcept {
+      ptr_.swap(other.ptr_);
+    }
+
+    /// @brief ADL-friendly swap overload
+    friend void swap(SharedRef &a, SharedRef &b) noexcept {
+      a.swap(b);
+    }
+
+    /// @brief Check if pointer is not null
+    explicit operator bool() const noexcept {
+      return static_cast<bool>(ptr_);
+    }
+
+    /// @brief Equality comparison
+    bool operator==(const SharedRef &other) const noexcept = default;
+
+    /// @brief Inequality comparison
+    bool operator!=(const SharedRef &other) const noexcept = default;
+
+    /// @brief Compare with raw shared_ptr
+    bool operator==(const std::shared_ptr<T> &other) const noexcept {
+      return ptr_ == other;
+    }
+
+    /// @brief Compare with raw shared_ptr
+    bool operator!=(const std::shared_ptr<T> &other) const noexcept {
+      return ptr_ != other;
+    }
+
+    /// @brief Less-than comparison for ordered containers
+    bool operator<(const SharedRef &other) const noexcept {
+      return ptr_.owner_before(other.ptr_);
+    }
+
+    /**
+     * @brief Indexing operator (only if T supports indexing)
+     * @tparam Index Index type
+     * @param i Index value
+     * @return Result of indexing T
+     */
+    template <typename Index>
+      requires requires(const T &t, Index i) {
+        { t[i] };
+      }
+    decltype(auto) operator[](Index i) const {
+      return (*ptr_)[i];
+    }
+
+    /**
+     * @brief Access the stored deleter (if known at compile time)
+     * @return Pointer to deleter object, or nullptr if not found
+     */
+    Deleter *get_deleter() const noexcept {
+      return std::get_deleter<Deleter>(ptr_);
+    }
+
+   private:
+    std::shared_ptr<T> ptr_;
+  };
+
+  /**
+   * @brief Deprecated alias for SharedRef.
+   *
+   * This alias exists for backward compatibility and will be removed in future
+   * versions.
+   *
+   * @tparam T The type of the managed object.
+   * @tparam Deleter The type of the deleter used to destroy the object.
+   *
+   * @deprecated Use SharedRef<T, Deleter> directly instead.
+   */
+  template <class T, class Deleter = std::default_delete<T>>
+    requires(not std::is_void_v<T>)
+  using StrictSharedPtr [[deprecated("Use SharedRef instead")]] =
+      SharedRef<T, Deleter>;
```

Comment:
```suggestion
  template <class T>
  using StrictSharedPtr [[deprecated("Use SharedRef instead")]] =
      SharedRef<T>;
```
I think that's enough here?